### PR TITLE
DEV-852 Activated functionality for delete resource button

### PR DESCRIPTION
**Pull request by nghianja**

### CODE REVIEW

File: src/main/java/sg/ncl/DataController.java [hunk 1]
```diff
@@ -322,6 +322,31 @@ public void getResource(@PathVariable String datasetId,
...
+            if (RestUtil.isError(response.getStatusCode())) {
+                MyErrorResource error = objectMapper.readValue(dataResponseBody, MyErrorResource.class);
+                ExceptionState exceptionState = ExceptionState.parseExceptionState(error.getError());
+
+                if (exceptionState == FORBIDDEN_EXCEPTION) {
+                    log.error("Removing of data resource forbidden.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                }
+            }
+        } catch (IOException e) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/main/java/sg/ncl/DataController.java [hunk 2]
```diff
@@ -322,6 +322,31 @@ public void getResource(@PathVariable String datasetId,
...
+
+        try {
+            if (RestUtil.isError(response.getStatusCode())) {
+                MyErrorResource error = objectMapper.readValue(dataResponseBody, MyErrorResource.class);
+                ExceptionState exceptionState = ExceptionState.parseExceptionState(error.getError());
+
+                if (exceptionState == FORBIDDEN_EXCEPTION) {
+                    log.error("Removing of data resource forbidden.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                }
```
Comment (ZHANGChunwang) [inline src/main/java/sg/ncl/DataController.java]: 1. During my local tests, I cannot delete a file and there is no error message shown on the UI. 
2. You only catch forbidden exception, which is actually unlikely to happen. 
3. You need to catch more exceptions such as data/resource not found exception, and default exception for all the other types of exceptions.
4. Not found could means that the data info is not available in the database, or the actual file is not found on the file system. Do we need to distinguish them?
5. What is more important is to show users some useful info when things went wrong.

File: src/main/java/sg/ncl/DataController.java [hunk 3]
```diff
@@ -322,6 +324,42 @@ public void getResource(@PathVariable String datasetId,
...
+                    log.error("Data resource not found for removing.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                } else if (exceptionState == DATA_RESOURCE_DELETE_EXCEPTION) {
+                    log.error("Error when removing data resource file.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                } else {
+                    log.warn("Unknown error.");
+                }
+            }
+        } catch (IOException e) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MAJOR 1 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [DataController.java#L149](https://github.com/nus-ncl/service-web/blob/8f3c77385b518b62af267b8b2388f866588142b3/src/main/java/sg/ncl/DataController.java#L149): Define a constant instead of duplicating this literal "Unknown error." 3 times. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1192)

File: src/main/java/sg/ncl/DataController.java [hunk 4]
```diff
@@ -322,6 +324,42 @@ public void getResource(@PathVariable String datasetId,
...
+                    log.error("Data resource not found for removing.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                } else if (exceptionState == DATA_RESOURCE_DELETE_EXCEPTION) {
+                    log.error("Error when removing data resource file.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                } else {
+                    log.warn("Unknown error for removing data resource.");
+                }
+            }
+        } catch (IOException e) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: src/main/java/sg/ncl/DataController.java [hunk 5]
```diff
@@ -280,6 +284,8 @@ public String getResources(Model model, @PathVariable String datasetId, HttpSess
 
             if (RestUtil.isError(responseEntity.getStatusCode())) {
                 throw new Exception();
+            } else if (body.equals("All finished.")) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MINOR Move the "All finished." string literal on the left side of this string comparison. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1132)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 2 issues
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: src/main/java/sg/ncl/DataController.java [hunk 6]
```diff
@@ -322,6 +336,44 @@ public void getResource(@PathVariable String datasetId,
...
+                } else if (exceptionState == DATA_RESOURCE_DELETE_EXCEPTION) {
+                    log.warn("Error when removing data resource file.");
+                    redirectAttributes.addFlashAttribute(MESSAGE_ATTRIBUTE, error.getMessage());
+                } else {
+                    log.warn("Unknown error for removing data resource.");
+                }
+            } else {
+                log.info("Data resource removed: {}", dataResponseBody);
+            }
+        } catch (IOException e) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MAJOR Either log or rethrow this exception. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1166)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 3 issues
* CRITICAL 1 critical
* MAJOR 1 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [DataController.java#L94](https://github.com/nus-ncl/service-web/blob/ba553435df63621a90eed7895adfeb672af11eab/src/main/java/sg/ncl/DataController.java#L94): The Cyclomatic Complexity of this method "validateContributeData" is 11 which is greater than 10 authorized. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 1 critical
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. CRITICAL [DataController.java#L260](https://github.com/nus-ncl/service-web/blob/a56403fd97c52e88c0d108043f01bac4ae0ba7a3/src/main/java/sg/ncl/DataController.java#L260): The Cyclomatic Complexity of this method "uploadChunk" is 15 which is greater than 10 authorized. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/sg/ncl/DataController.java [hunk 7]
```diff
@@ -279,7 +295,12 @@ public String getResources(Model model, @PathVariable String datasetId, HttpSess
             String body = responseEntity.getBody().toString();
 
             if (RestUtil.isError(responseEntity.getStatusCode())) {
-                throw new Exception();
+                MyErrorResource error = objectMapper.readValue(body, MyErrorResource.class);
+                ExceptionState exceptionState = ExceptionState.parseExceptionState(error.getError());
+
+                return getStringResponseEntity(exceptionState);
+            } else if (body.equals("All finished.")) {
```
Comment (ChrisZhong) [inline src/main/java/sg/ncl/DataController.java]: MINOR Move the "All finished." string literal on the left side of this string comparison. [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1132)

**Comment by ChrisZhong** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

**Pull request merged**